pkg/scheduler/backend/api_dispatcher/call_controller.go

```
@@ -0,0 +1,300 @@
+	apiCallRelevances fwk.APICallRelevances
+	// callIDCounter is a monotonically increasing counter used to assign a unique callID
+	// to each queuedAPICall added to the callController.
+	callIDCounter int
+
+	// apiCalls stores details of API calls for objects that are currently enqueued or in-flight.
+	// Must be used under lock.
+	apiCalls map[types.UID]*queuedAPICall
+	// callsQueue is a FIFO queue that stores the object UIDs of (pending) API calls waiting to be executed.
+	// Must be used under lock.
+	callsQueue buffer.Ring[types.UID]
+	// inFlightEntities stores object UIDs of API calls that are currently in-flight
+	// (i.e., their execution goroutine has been dispatched).
+	// Must be used under lock.
+	inFlightEntities sets.Set[types.UID]
+}
+
+// newCallController returns a new callController object.
+func newCallController(apiCallRelevances fwk.APICallRelevances) *callController {
+	q := callController{
+		apiCallRelevances: apiCallRelevances,
+		apiCalls:          make(map[types.UID]*queuedAPICall),
+		inFlightEntities:  sets.New[types.UID](),
+	}
+	q.cond = sync.NewCond(&q.lock)
+
+	return &q
+}
+
+// isLessRelevant returns true if newCall is less relevant than oldCall.
+func (cc *callController) isLessRelevant(oldCall, newCall fwk.APICallType) bool {
+	return cc.apiCallRelevances[newCall] < cc.apiCallRelevances[oldCall]
+}
+
+// reconcile compares a new API call with an existing one for the same object.
+// Based on their relevance and type, it determines whether to skip the new call,
+// overwrite the old one, or merge them.
+// Must be called under cc.lock.
+func (cc *callController) reconcile(oldAPICall, apiCall *queuedAPICall) error {
+	if cc.isLessRelevant(oldAPICall.CallType(), apiCall.CallType()) {
+		// The new API call is less relevant than the existing one, so skip it.
+		err := fmt.Errorf("a more relevant call is already enqueued for this object: %w", fwk.ErrCallSkipped)
+		apiCall.sendOnFinish(err)
```

Question: how should users (= who made this API call) react to this error?
If they can just ignore the errors, then we can just log it here, and no need to send err to the channel?

---

I don't want to just return nil (== everything went fine) if the call was overwritten or skipped. In that case, the newer call wasn't even executed yet, so the logic that would handle that result, might have to behave differently.

For example, cache processing successful update, might just store the update's result. But, if it receives an overwrite, it would mean that the cache already started tracking the more relevant call (e.g. binding) and should drop the previous one (e.g. status update).

Similarly, skipped should mean that the update won't be tracked, because the more important call is already enqueued.